topdown/topdown_partial_test.go

```diff
@@ -2168,8 +2217,91 @@ func TestTopDownPartialEval(t *testing.T) {
 			wantSupport: []string{
 				`package partial.test
 
-				p = true { __local2__1 = input; __local2__1 = __local0__2; equal(__local0__2, 1, __local1__2); y2 = __local1__2; y2 }
-				`,
+				p = true { __local2__1 = input; __local2__1 = __local0__2; equal(__local0__2, 1, __local1__2); y2 = __local1__2; y2 }`,
+			},
+		},
+		{
+			note:    "shallow inlining: function with unknowns in rule body",
+			query:   "data.test.f(1, x)",
+			shallow: true,
+			modules: []string{
+				`package test
+				f(x) = true { input.x = x }
+				f(x) = false { input.y = x }`,
+			},
+			wantQueries: []string{`data.partial.test.f(1, x)`},
+			wantSupport: []string{
+				`package partial.test
+				f(__local0__2) = true { input.x = __local0__2 }
+				f(__local1__1) = false { input.y = __local1__1 }`,
+			},
+		},
+		{
+			note:    "shallow inlining: functions with no unknowns in rule body or output",
+			query:   "data.test.f(1, y)",
+			shallow: true,
+			modules: []string{
+				`package test
+				f(x) = true { x > 1 }
+				f(x) = false { x <= 0 }
+				f(x) = "meow" { false }`,
+			},
+			wantQueries: []string{},
+		},
+		{
+			note:    "shallow inlining: functions with multiple args, no unknowns",
+			query:   "data.test.f(1, [1,2,3], y)",
+			shallow: true,
+			modules: []string{
+				`package test
+				f(x, y) = true { x > 1 }
+				f(x, y) = false {
+					x <= 0
+					count(y) == 3
+				}`,
+			},
+			wantQueries: []string{},
+		},
+		{
+			note:    "shallow inlining: functions that are always undefined",
+			query:   "data.test.f(1, y)",
+			shallow: true,
+			modules: []string{
+				`package test
+				f(x) = "uhm" { false }
+				f(x) = "like" { false }
+				f(x) = "whatever" { false }`,
+			},
+			wantQueries: []string{},
+		},
+		{
+			note:    "shallow inlining: functions with non-var arguments",
+			query:   "data.test.f(1, y)",
+			shallow: true,
+			modules: []string{
+				`package test
+				f(true) = true
+				f(x) = false { x != true }`,
+			},
+			wantQueries: []string{`y = false`},
+		},
+		{
+			note:    "shallow inlining: functions with unknown call-site arguments",
+			query:   "input = x; data.test.f([1, x])",
+			shallow: true,
+			modules: []string{
+				`package test
+				f([x, y]) {
+				  z = 7
+				  x > (y+z)
+				}
+				f(_) = false`, // two bodies to trigger shallow inlining
```

Comment:
Q: Why are two bodies required to trigger shallow inlining? I would not have expected that.

Reply:
It's an optimisation we could discuss: if there's only one definition, inlining has no unwanted side effects, i.e. cross products. Or so I thought...

Reply:
There could still be an unwanted cross-product if we unroll a loop inside a single function body. Here's a contrived example:

```rego
p[x]; f(x)
``` 

```rego
f(x) { q[y]; input = x; input = y }
```

With the current implementation would generate p*q queries.